lib/src/clojure_lsp/feature/diagnostics.clj

```diff
@@ -213,6 +214,39 @@
 (defn publish-empty-diagnostics! [uris components]
   (publish-all-diagnostics!* components (map empty-diagnostics-of-uri uris)))
 
+(defn ^:private namespace-alias->finding [element inconsistencies kondo-config]
+  {:uri (:uri element)
+   :row (:name-row element)
+   :col (:name-col element)
+   :end-row (:name-end-row element)
+   :end-col (:name-end-col element)
+   :level (-> kondo-config :linters :clojure-lsp/uniform-aliasing :level)
+   :message (format "Different aliases %s found for %s"
+                    (get inconsistencies (:to element))
+                    (:to element))
+   :type :clojure-lsp/uniform-aliasing})
+
+(defn ^:private uniform-aliasing [narrowed-db project-db kondo-config]
+  (let [kondo-config (update-in kondo-config [:linters :clojure-lsp/uniform-aliasing :level] #(or % :off))]
+    (when-not (identical? :off (-> kondo-config :linters :clojure-lsp/uniform-aliasing :level))
+      (let [exclude-aliases-config (-> kondo-config :linters :clojure-lsp/uniform-aliasing :exclude-aliases set)
+            exclude-aliases (conj exclude-aliases-config nil) ;; nil here means an unaliased require
+            dep-graph (:dep-graph narrowed-db)
+            inconsistencies (into {} (for [[ns dep-graph-item] dep-graph
+                                           :let [aliases-assigned (-> dep-graph-item :aliases-breakdown :internal keys set (set/difference exclude-aliases))]
+                                           :when (> (count aliases-assigned) 1)]
+                                       [ns aliases-assigned]))]
+        (for [{dependents :dependents} (map dep-graph (keys inconsistencies))
```

Comment:
would be nice during the profiling to check how faster is a version using transducers vs these fors

Reply:
Indeed, I've managed to get a `project-db` from [metabase](https://github.com/metabase/metabase) and run a few manual experiments (nothing too scientific).
```clojure
(def project-db (time (with-open [r (clojure.java.io/reader "/Users/andre.camargo/dev/nu/stormshield/project-db.2024-02-12T12:18:44.782017.edn")]
                                   (clojure.edn/read (java.io.PushbackReader. r)))))
"Elapsed time: 3635.460292 msecs"
#'clojure-lsp.feature.diagnostics/project-db
clj꞉clojure-lsp.feature.diagnostics꞉> 
(keys project-db)
(:findings
 :dep-graph
 :classpath
 :project-hash
 :client-settings
 :settings
 :project-root-uri
 :env
 :kondo-config
 :kondo-config-hash
 :dependency-scheme
 :documents
 :analysis
 :project-settings
 :analysis-checksums
 :api?
 :project-analysis-type
 :full-scan-analysis-startup
 :client-capabilities
 :classpath-settings
 :force-settings)
clj꞉clojure-lsp.feature.diagnostics꞉> 
(count (:dep-graph project-db))
7022
clj꞉clojure-lsp.feature.diagnostics꞉> 
(count (:analysis project-db))
2708
clj꞉clojure-lsp.feature.diagnostics꞉> 
(def kondo-config
       {:linters {:clojure-lsp/different-aliases {:level :error}}})
#'clojure-lsp.feature.diagnostics/kondo-config
clj꞉clojure-lsp.feature.diagnostics꞉> 
(defn ^:private different-aliases [narrowed-db project-db kondo-config]
        (let [kondo-config (update-in kondo-config [:linters :clojure-lsp/different-aliases :level] #(or % :off))]
             (when-not (identical? :off (-> kondo-config :linters :clojure-lsp/different-aliases :level))
                       (let [exclude-aliases-config (-> kondo-config :linters :clojure-lsp/different-aliases :exclude-aliases set)
                             exclude-aliases (conj exclude-aliases-config nil) ;; nil here means an unaliased require
                             dep-graph (:dep-graph narrowed-db)
                             inconsistencies (reduce (fn [m [ns dep-graph-item]]
                                                         (let [aliases-assigned (-> dep-graph-item :aliases-breakdown :internal keys set (set/difference exclude-aliases))]
                                                              (if (> (count aliases-assigned) 1)
                                                                  (assoc m ns aliases-assigned)
                                                                  m)))
                                                     {}
                                                     dep-graph)]
                            (for [{dependents :dependents} (map dep-graph (keys inconsistencies))
                                  [namespace _] dependents
                                  :let [dep-graph-item (get dep-graph namespace)]
                                  uri (:uris dep-graph-item)
                                  :let [var-definition (-> project-db :analysis (get uri))]
                                  namespace-alias (:namespace-alias var-definition)
                                  :when (let [{:keys [alias to]} namespace-alias]
                                             (and (contains? inconsistencies to)
                                                  (some #{alias} (get inconsistencies to))))]
                                 (namespace-alias->finding namespace-alias inconsistencies kondo-config))))))
#'clojure-lsp.feature.diagnostics/different-aliases
clj꞉clojure-lsp.feature.diagnostics꞉> 
(count (time (different-aliases project-db project-db kondo-config)))
"Elapsed time: 6.078334 msecs"
59706
clj꞉clojure-lsp.feature.diagnostics꞉> 
(count (time (different-aliases project-db project-db kondo-config)))
"Elapsed time: 5.314 msecs"
59706
clj꞉clojure-lsp.feature.diagnostics꞉> 
(count (time (different-aliases project-db project-db kondo-config)))
"Elapsed time: 4.421125 msecs"
59706
clj꞉clojure-lsp.feature.diagnostics꞉> 
(defn ^:private xf-different-aliases [narrowed-db project-db kondo-config]
        (let [kondo-config (update-in kondo-config [:linters :clojure-lsp/different-aliases :level] #(or % :off))]
             (when-not (identical? :off (-> kondo-config :linters :clojure-lsp/different-aliases :level))
                       (let [exclude-aliases-config (-> kondo-config :linters :clojure-lsp/different-aliases :exclude-aliases set)
                             exclude-aliases (conj exclude-aliases-config nil) ;; nil here means an unaliased require
                             dep-graph (:dep-graph narrowed-db)
                             inconsistencies (reduce (fn [m [ns dep-graph-item]]
                                                         (let [aliases-assigned (-> dep-graph-item :aliases-breakdown :internal keys set (set/difference exclude-aliases))]
                                                              (if (> (count aliases-assigned) 1)
                                                                  (assoc m ns aliases-assigned)
                                                                  m)))
                                                     {}
                                                     dep-graph)]
                            (for [{dependents :dependents} (map dep-graph (keys inconsistencies))
                                  [namespace _] dependents
                                  :let [dep-graph-item (get dep-graph namespace)]
                                  uri (:uris dep-graph-item)
                                  :let [var-definition (-> project-db :analysis (get uri))]
                                  namespace-alias (:namespace-alias var-definition)
                                  :when (let [{:keys [alias to]} namespace-alias]
                                             (and (contains? inconsistencies to)
                                                  (some #{alias} (get inconsistencies to))))]
                                 (namespace-alias->finding namespace-alias inconsistencies kondo-config))))))
#'clojure-lsp.feature.diagnostics/xf-different-aliases
clj꞉clojure-lsp.feature.diagnostics꞉> 
(count (time (xf-different-aliases project-db project-db kondo-config)))
"Elapsed time: 5.214042 msecs"
59706
clj꞉clojure-lsp.feature.diagnostics꞉> 
(count (time (xf-different-aliases project-db project-db kondo-config)))
"Elapsed time: 3.174791 msecs"
59706
clj꞉clojure-lsp.feature.diagnostics꞉> 
(count (time (xf-different-aliases project-db project-db kondo-config)))
"Elapsed time: 4.206125 msecs"
59706
```
Using list comprehensions the median is 5.314 msecs while using transducers is 3.174791 msecs.

Reply:
Thanks, I think we can go without transducers for clarity as the performance looks good, but I'm not sure the `project-db` you are using is correct, 1466 elements for metabase project-db looks really low

Reply:
I've managed to connect to Clojure-LSP REPL server on metabase and...
```clojure
(def project-db @clojure-lsp.db/db*)
#'clojure-lsp.feature.diagnostics/project-db
clj꞉clojure-lsp.feature.diagnostics꞉> 
(keys project-db)
(:findings
 :dep-graph
 :classpath
 :settings-auto-refresh?
 :project-hash
 :clj-depend-violations
 :client-settings
 :settings
 :project-root-uri
 :port
 :kondo-config
 :kondo-config-hash
 :dependency-scheme
 :documents
 :analysis
 :project-settings
 :analysis-checksums
 :project-analysis-type
 :log-path
 :clojuredocs
 :client-capabilities
 :dev
 :force-settings
 :stubs-generation-namespaces)
clj꞉clojure-lsp.feature.diagnostics꞉> 
(count (:dep-graph project-db))
2939
clj꞉clojure-lsp.feature.diagnostics꞉> 
(count (:analysis project-db))
95775
clj꞉clojure-lsp.feature.diagnostics꞉> 
(def kondo-config
  {:linters {:clojure-lsp/different-aliases {:level :error}}})
#'clojure-lsp.feature.diagnostics/kondo-config
clj꞉clojure-lsp.feature.diagnostics꞉> 
(defn ^:private different-aliases [narrowed-db project-db kondo-config]
  (let [kondo-config (update-in kondo-config [:linters :clojure-lsp/different-aliases :level] #(or % :off))]
    (when-not (identical? :off (-> kondo-config :linters :clojure-lsp/different-aliases :level))
      (let [exclude-aliases-config (-> kondo-config :linters :clojure-lsp/different-aliases :exclude-aliases set)
            exclude-aliases (conj exclude-aliases-config nil) ;; nil here means an unaliased require
            dep-graph (:dep-graph narrowed-db)
            inconsistencies (reduce (fn [m [ns dep-graph-item]]
                                      (let [aliases-assigned (-> dep-graph-item :aliases-breakdown :internal keys set (set/difference exclude-aliases))]
                                        (if (> (count aliases-assigned) 1)
                                          (assoc m ns aliases-assigned)
                                          m)))
                                    {}
                                    dep-graph)]
        (for [{dependents :dependents} (map dep-graph (keys inconsistencies))
              [namespace _] dependents
              :let [dep-graph-item (get dep-graph namespace)]
              uri (:uris dep-graph-item)
              :let [var-definition (-> project-db :analysis (get uri))]
              namespace-alias (:namespace-alias var-definition)
              :when (let [{:keys [alias to]} namespace-alias]
                      (and (contains? inconsistencies to)
                           (some #{alias} (get inconsistencies to))))]
          (namespace-alias->finding namespace-alias inconsistencies kondo-config))))))
#'clojure-lsp.feature.diagnostics/different-aliases
clj꞉clojure-lsp.feature.diagnostics꞉> 
(count (time (different-aliases project-db project-db kondo-config)))
"Elapsed time: 6.547375 msecs"
4132
clj꞉clojure-lsp.feature.diagnostics꞉> 
(count (time (different-aliases project-db project-db kondo-config)))
"Elapsed time: 2.233042 msecs"
4132
clj꞉clojure-lsp.feature.diagnostics꞉> 
(count (time (different-aliases project-db project-db kondo-config)))
"Elapsed time: 2.721 msecs"
4132
clj꞉clojure-lsp.feature.diagnostics꞉> 
(count (time (different-aliases project-db project-db kondo-config)))
"Elapsed time: 2.140375 msecs"
4132
clj꞉clojure-lsp.feature.diagnostics꞉> 
(count (time (different-aliases project-db project-db kondo-config)))
"Elapsed time: 3.10175 msecs"
4132
clj꞉clojure-lsp.feature.diagnostics꞉> 
(defn ^:private xf-different-aliases [narrowed-db project-db kondo-config]
  (let [kondo-config (update-in kondo-config [:linters :clojure-lsp/different-aliases :level] #(or % :off))]
    (when-not (identical? :off (-> kondo-config :linters :clojure-lsp/different-aliases :level))
      (let [exclude-aliases-config (-> kondo-config :linters :clojure-lsp/different-aliases :exclude-aliases set)
            exclude-aliases (conj exclude-aliases-config nil) ;; nil here means an unaliased require
            dep-graph (:dep-graph narrowed-db)
            inconsistencies (reduce (fn [m [ns dep-graph-item]]
                                      (let [aliases-assigned (-> dep-graph-item :aliases-breakdown :internal keys set (set/difference exclude-aliases))]
                                        (if (> (count aliases-assigned) 1)
                                          (assoc m ns aliases-assigned)
                                          m)))
                                    {}
                                    dep-graph)]
        (for [{dependents :dependents} (map dep-graph (keys inconsistencies))
              [namespace _] dependents
              :let [dep-graph-item (get dep-graph namespace)]
              uri (:uris dep-graph-item)
              :let [var-definition (-> project-db :analysis (get uri))]
              namespace-alias (:namespace-alias var-definition)
              :when (let [{:keys [alias to]} namespace-alias]
                      (and (contains? inconsistencies to)
                           (some #{alias} (get inconsistencies to))))]
          (namespace-alias->finding namespace-alias inconsistencies kondo-config))))))
#'clojure-lsp.feature.diagnostics/xf-different-aliases
clj꞉clojure-lsp.feature.diagnostics꞉> 
(count (time (xf-different-aliases project-db project-db kondo-config)))
"Elapsed time: 3.405041 msecs"
4132
clj꞉clojure-lsp.feature.diagnostics꞉> 
(count (time (xf-different-aliases project-db project-db kondo-config)))
"Elapsed time: 1.674667 msecs"
4132
clj꞉clojure-lsp.feature.diagnostics꞉> 
(count (time (xf-different-aliases project-db project-db kondo-config)))
"Elapsed time: 3.07625 msecs"
4132
clj꞉clojure-lsp.feature.diagnostics꞉> 
(count (time (xf-different-aliases project-db project-db kondo-config)))
"Elapsed time: 4.136083 msecs"
4132
clj꞉clojure-lsp.feature.diagnostics꞉> 
(count (time (xf-different-aliases project-db project-db kondo-config)))
"Elapsed time: 2.300333 msecs"
4132
clj꞉clojure-lsp.feature.diagnostics꞉> 
```
list comprehension median is 2.721 msecs and transducer median is 3.07625 msecs